pkg/function/create_csi_snapshot.go

```diff
@@ -0,0 +1,129 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package function
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/util/rand"
+
+	kanister "github.com/kanisterio/kanister/pkg"
+	"github.com/kanisterio/kanister/pkg/kube"
+	"github.com/kanisterio/kanister/pkg/kube/snapshot"
+	"github.com/kanisterio/kanister/pkg/param"
+)
+
+func init() {
+	_ = kanister.Register(&createCSISnapshotFunc{})
+}
+
+var (
+	_ kanister.Func = (*createCSISnapshotFunc)(nil)
+)
+
+const (
+	// CreateCSIVolumeSnapshotFuncName gives the name of the function
+	CreateCSISnapshotFuncName = "CreateCSISnapshot"
+	// CreateCSISnapshotNameArg provides name of the new VolumeSnapshot
+	CreateCSISnapshotNameArg = "name"
+	// CreateCSISnapshotPVCNameArg gives the name of the captured PVC
+	CreateCSISnapshotPVCNameArg = "pvc"
+	// CreateCSISnapshotNamespaceArg mentions the namespace of the captured PVC
+	CreateCSISnapshotNamespaceArg = "namespace"
+	// CreateCSISnapshotSnapshotClassArg specifies the name of the VolumeSnapshotClass
+	CreateCSISnapshotSnapshotClassArg = "snapshotClass"
+	// CreateCSISnapshotLabelsArg has labels that are to be added to the new VolumeSnapshot
+	CreateCSISnapshotLabelsArg = "labels"
+	// CreateCSISnapshotRestoreSizeArg gives the storage size required for PV/PVC restoration
+	CreateCSISnapshotRestoreSizeArg = "restoreSize"
+	// CreateCSISnapshotSnapshotContentNameArg provides the name of dynamically provisioned VolumeSnapshotContent
+	CreateCSISnapshotSnapshotContentNameArg = "snapshotContent"
+)
+
+type createCSISnapshotFunc struct{}
+
+func (*createCSISnapshotFunc) Name() string {
+	return CreateCSISnapshotFuncName
+}
+
+func (*createCSISnapshotFunc) Exec(ctx context.Context, tp param.TemplateParams, args map[string]interface{}) (map[string]interface{}, error) {
+	var snapshotClass string
+	var labels map[string]string
+	var name, pvc, namespace string
+	if err := Arg(args, CreateCSISnapshotPVCNameArg, &pvc); err != nil {
+		return nil, err
+	}
+	if err := Arg(args, CreateCSISnapshotNamespaceArg, &namespace); err != nil {
+		return nil, err
+	}
+	if err := Arg(args, CreateCSISnapshotSnapshotClassArg, &snapshotClass); err != nil {
+		return nil, err
+	}
+	if err := OptArg(args, CreateCSISnapshotNameArg, &name, defaultSnapshotName(pvc, 20)); err != nil {
+		return nil, err
+	}
+	if err := OptArg(args, CreateCSISnapshotLabelsArg, &labels, map[string]string{}); err != nil {
+		return nil, err
+	}
+
+	kubeCli, err := kube.NewClient()
+	if err != nil {
+		return nil, err
+	}
+	dynCli, err := kube.NewDynamicClient()
+	if err != nil {
+		return nil, err
+	}
+	snapshotter, err := snapshot.NewSnapshotter(kubeCli, dynCli)
+	if err != nil {
+		if errors.Is(context.DeadlineExceeded, err) {
+			return nil, errors.New("SnapshotContent not provisioned within given timeout. Please check if CSI driver is installed correctly and supports VolumeSnapshot feature")
```

Comment:
```suggestion
			return nil, errors.Wrap(err, "SnapshotContent not provisioned within given timeout. Please check if CSI driver is installed correctly and supports VolumeSnapshot feature")
```
@shlokchaudhari9 please add a message to the existing error.

Reply:
Done. There is `errors.Wrap()` supported so the syntax would be a little different.

Reply:
Its there in pkg `github.com/pkg/errors`
ref: https://github.com/kanisterio/kanister/blob/master/pkg/kopia/snapshot.go#L47